vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1179 @@
+                node_* temp = nullptr;
+                if (existing_nodes.find(other_loc) != existing_nodes.end()) {
+                    temp = existing_nodes[other_loc];
+                } else if (nodes.find(other_loc) != nodes.end()) {
+                    temp = new node_(std::get<0>(other_loc), std::get<1>(other_loc));
+                    temp->segment_id = node_tile_to_segment_[std::make_tuple(node_id, loc_to_tile_[other_loc])];
+                    nodes.erase(other_loc);
+                    existing_nodes.emplace(other_loc, temp);
+                    builder.emplace(1, -1 * visited_tiles, temp);
+                    visited_tiles++;
+                } else if (empties.find(other_loc) != empties.end()) {
+                    temp = new node_(std::get<0>(other_loc), std::get<1>(other_loc));
+                    temp->segment_id = -1;
+                    empties.erase(other_loc);
+                    existing_nodes.emplace(other_loc, temp);
+                    builder.emplace(0, -1 * visited_tiles, temp);
+                    visited_tiles++;
+                }
+                vertex->links[i] = temp;
+            }
+        }
+
+        if (!nodes.empty())
+            vpr_throw(VPR_ERROR_ARCH, get_arch_file_name(), __LINE__, "Node graph could not be created\n");
+
+        while (!builder.empty()) {
+            node_* vertex;
+            std::tie(std::ignore, std::ignore, vertex) = builder.top();
+            builder.pop();
+            location loc = vertex->loc;
+            for (int i = 0; i < 4; ++i) {
+                location offset = offsets[i];
+                location other_loc{std::get<0>(loc) + std::get<0>(offset), std::get<1>(loc) + std::get<1>(offset)};
+                node_* temp = nullptr;
+                if (existing_nodes.find(other_loc) != existing_nodes.end()) {
+                    temp = existing_nodes[other_loc];
+                    temp->links[(i + 2) % 4] = nullptr;
+                }
+            }
+            existing_nodes.erase(vertex->loc);
+            delete (vertex);
+        }
+
+        for (auto& node : existing_nodes) {
+            if (used_by_pip_.find(std::make_tuple(node_id, node.second->loc)) != used_by_pip_.end()) {
+                node.second->on_route = true;
+                root_node = node.second;
+            }
+            if (used_by_pin_.find(std::make_tuple(node_id, node.second->loc)) != used_by_pin_.end()) {
+                node.second->has_pins = true;
+                node.second->on_route = true;
+                root_node = node.second;
+            }
+        }
+        return root_node;
+    }
+
+    /*
+     * Removes dangling nodes from a graph represented by the root node.
+     * Dangling nodes are nodes that do not connect to a pin, a pip or other non-dangling node.
+     */
+    int reduce_graph_and_count_nodes(node_* root) {
+        int cnt = 0;
```

Comment:
Rename this with a more descriptive name

Reply:
I hope this is good enough